File: Example/ObjcExample/CollectionView/CollectionViewController.m
//
//  ViewController.m
//  ObjCSkeletonDemo
//
//  Created by Mikhail Zoline on 8/13/20.
//  Copyright © 2020 MZ. All rights reserved.
//
@import SkeletonView;
#import "ObjCExample-Swift.h"
#import "CollectionViewController.h"

@interface CollectionViewController ()
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@property (weak, nonatomic) IBOutlet UIImageView *avatarImage;
@property (weak, nonatomic) IBOutlet UIView *colorSelectedView;
@property (weak, nonatomic) IBOutlet UISwitch *switchAnimated;
@property (weak, nonatomic) IBOutlet UISegmentedControl *skeletonTypeSelector;
@property (weak, nonatomic) IBOutlet UIButton *showOrHideSkeletonButton;
@property (weak, nonatomic) IBOutlet UILabel *transitionDurationLabel;
@property (weak, nonatomic) IBOutlet UIStepper *transitionDurationStepper;

@property (nonatomic, readonly, assign) BOOL type;
@end

typedef CAAnimation* (^SkeletonLayerAnimation)(CALayer *);

@implementation CollectionViewController

- (void)configureCollectionView {
    _collectionView.isSkeletonable = true;
    _collectionView.backgroundColor = [UIColor clearColor];
    _collectionView.showsHorizontalScrollIndicator = false;
    _collectionView.showsVerticalScrollIndicator = false;
    _collectionView.delegate = self;
    _collectionView.dataSource = self;
    [_collectionView registerClass:[CollectionViewCell class] forCellWithReuseIdentifier:@"CollectionViewCell"];
}

- (void)configureAvatarImage {
    _avatarImage.layer.cornerRadius = self.avatarImage.frame.size.width/2;
    _avatarImage.layer.masksToBounds = true;
}

- (void)configureColorSelectedView {
    _colorSelectedView.layer.cornerRadius = 5;
    _colorSelectedView.layer.masksToBounds = true;
    _colorSelectedView.backgroundColor = SkeletonAppearance.shared.tintColor;
    _colorSelectedView.tintColor = [Colors color:16];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _transitionDurationStepper.value = 0.25;
    [self configureCollectionView];
    [self configureColorSelectedView];
    [self configureAvatarImage];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    [self.collectionView prepareSkeletonWithCompletion:^(BOOL done) {
        [self showSkeleton];
    }];
}

- (IBAction)changeAnimated:(id)sender {
    if (_switchAnimated.isOn) {
        [self.view startSkeletonAnimation:NULL];
    } else {
        [self.view stopSkeletonAnimation];
    }
}

- (IBAction)btnChangeColorTouchUpInside:(id)sender {
    [self showAlertPicker];
}

- (IBAction)showOrHideSkeleton:(id)sender {
    [_showOrHideSkeletonButton setTitle: self.view.isSkeletonActive ? @"Show skeleton" : @"Hide skeleton" forState:UIControlStateNormal];
    self.view.isSkeletonActive ? [self hideSkeleton] : [self showSkeleton];
}

- (IBAction)transitionDurationStepperAction:(id)sender {
    _transitionDurationLabel.text = [NSString stringWithFormat:@"Fade duration: %.2f s", _transitionDurationStepper.value];
    if (self.view.isSkeletonActive) { [self showSkeleton]; }
}

- (IBAction)changeSkeletonType:(id)sender {
    [self refreshSkeleton];
}

- (BOOL)type {
    return _skeletonTypeSelector.selectedSegmentIndex == 0 ? SkeletonType.solid: SkeletonType.gradient;
}

- (void)showSkeleton {
    [self refreshSkeleton];
}

- (void)hideSkeleton {
    [self.view hideSkeletonWithReloadDataAfter:
     true transition:
     [[SkeletonTransitionStyle alloc]init:
      _transitionDurationStepper.value]];
}

- (void)refreshSkeleton {
    if (self.view.isSkeletonActive) {
        SkeletonTransitionStyle *style = [[SkeletonTransitionStyle alloc]init: 0];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideSkeletonWithReloadDataAfter:true transition: style];
        });
    }
    
    [NSThread sleepForTimeInterval:0.1f];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if ( self.type == SkeletonType.gradient) {
            [self showGradientSkeleton];
        } else {
            [self showSolidSkeleton ];
        }
    });
}

- (void)showSolidSkeleton {
    if (_switchAnimated.isOn) {
        [self.view showAnimatedSkeletonUsingColor:
         _colorSelectedView.backgroundColor animation:
         NULL transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    } else {
        [self.view showSkeletonUsingColor:
         _colorSelectedView.backgroundColor transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    }
}

- (void)showGradientSkeleton {
    SkeletonGradient* gradient = [[SkeletonGradient alloc] initWithBaseColor:_colorSelectedView.backgroundColor secondaryColor:_collectionView.tintColor];
    
    if (_switchAnimated.isOn) {
        [self.view showAnimatedGradientSkeletonUsingGradient:
         gradient animation:
         NULL transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    } else {
        [self.view showGradientSkeletonUsingGradient:
         gradient transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    }
}

- (void)showAlertPicker {
    UIAlertController* alertView =
    [UIAlertController alertControllerWithTitle:
     @"Select color" message:@"\n\n\n\n\n\n" preferredStyle:
     UIAlertControllerStyleAlert];
    
    UIPickerView* pickerView =
    [[UIPickerView alloc] initWithFrame:
     CGRectMake(0.0, 50.0, 260.0, 120.0)];
    [pickerView setDataSource:self];
    [pickerView setDelegate:self];
    [alertView.view addSubview:pickerView];
    
    UIAlertAction* action =
    [UIAlertAction actionWithTitle:
     @"Ok" style:
     UIAlertActionStyleDefault handler:
     ^(UIAlertAction * _Nonnull action) {
        __weak CollectionViewController* weakSelf = self;
        __weak UIPickerView* weakPicker = pickerView;
        long row = [weakPicker selectedRowInComponent:0];
        weakSelf.colorSelectedView.backgroundColor = [Colors color:row];
        [weakSelf refreshSkeleton];
    }];
    [alertView addAction:action];
    
    UIAlertAction* cancelAction =
    [UIAlertAction actionWithTitle:
     @"Cancel" style:
     UIAlertActionStyleCancel handler:
     NULL];
    [alertView addAction:cancelAction];
    [self presentViewController:
     alertView animated:
     false completion:
     NULL];
}

// MARK: - UIPickerViewDelegate, UIPickerViewDataSource
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return [Colors count];
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return [Colors title:row] ;
}

// MARK: - UICollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return CGSizeMake(self.view.frame.size.width/3 - 10, self.view.frame.size.width/3 - 10);
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return 5;
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return 5;
}
- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section {
    return UIEdgeInsetsMake(0, 7, 0, 7);
}

// MARK: - SkeletonCollectionViewDataSource
- (NSString *)collectionSkeletonView:(UICollectionView *)skeletonView supplementaryViewIdentifierOfKind:(NSString *)supplementaryViewIdentifierOfKind at:(NSIndexPath *)indexPath {
    return nil;
}
- (NSInteger)numSectionsIn:(UICollectionView *)collectionSkeletonView {
    return 1;
}

- (NSString *)collectionSkeletonView:(UICollectionView *)skeletonView cellIdentifierForItemAt:(NSIndexPath *)indexPath {
    return @"CollectionViewCell";
}

- (NSInteger)collectionSkeletonView:(UICollectionView *)skeletonView numberOfItemsInSection:(NSInteger)section {
    return 10;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    // this code is not called
    // since SkeletonCollectionViewDataSource does its own imp of this
    //
    CollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"CollectionViewCell" forIndexPath:indexPath];
    if(!cell) {
        cell = [CollectionViewCell init];
    }
    return cell;
}


@end

File: Example/SwiftExample/CollectionView/CollectionViewController.swift
// CollectionViewController.swift
//

import UIKit
import SkeletonView

class CollectionViewController: UIViewController {
    @IBOutlet weak var collectionView: UICollectionView! {
        didSet {
            collectionView.isSkeletonable = true
            collectionView.backgroundColor = .clear
            collectionView.showsHorizontalScrollIndicator = false
            collectionView.showsVerticalScrollIndicator = false
            
            collectionView.dataSource = self
            collectionView.delegate = self
            
            collectionView.register(CollectionViewCell.self, forCellWithReuseIdentifier: "CollectionViewCell")
        }
    }
    
    @IBOutlet weak var avatarImage: UIImageView! {
        didSet {
            avatarImage.layer.cornerRadius = avatarImage.frame.width/2
            avatarImage.layer.masksToBounds = true
        }
    }
    
    @IBOutlet weak var colorSelectedView: UIView! {
        didSet {
            colorSelectedView.layer.cornerRadius = 5
            colorSelectedView.layer.masksToBounds = true
            colorSelectedView.backgroundColor = Colors.color(0)
            colorSelectedView.tintColor = Colors.color(1)
        }
    }
    
    @IBOutlet weak var switchAnimated: UISwitch!
    @IBOutlet weak var skeletonTypeSelector: UISegmentedControl!
    @IBOutlet weak var showOrHideSkeletonButton: UIButton!
    @IBOutlet weak var transitionDurationLabel: UILabel!
    @IBOutlet weak var transitionDurationStepper: UIStepper!
    
    var secondaryColor: UIColor = SkeletonAppearance.shared.tintColor
    
    var type: SkeletonType {
        return SkeletonType(solidOrGradient: skeletonTypeSelector.selectedSegmentIndex == 0 ?  SkeletonType.solid : SkeletonType.gradient)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        transitionDurationStepper.value = 0.25
        collectionView.prepareSkeleton { (done) in
            self.view.showAnimatedGradientSkeleton()
        }
    }
    
    class var defaultAnim: SkeletonLayerAnimation {
        get {
            let animBuilder  = GradientDirection(direction: GradientDirection.bottomRightTopLeft)
            return animBuilder.slidingAnimation()
        }
    }
    
    @IBAction func changeAnimated(_ sender: Any) {
        if switchAnimated.isOn {
            view.startSkeletonAnimation( CollectionViewController.defaultAnim)
        } else {
            view.stopSkeletonAnimation()
        }
    }
    
    @IBAction func changeSkeletonType(_ sender: Any) {
        if self.view.isSkeletonActive {
            self.hideSkeleton()
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                self.showSkeleton()
            }
        }
    }
    
    @IBAction func btnChangeColorTouchUpInside(_ sender: Any) {
        showAlertPicker()
    }
    
    @IBAction func showOrHideSkeleton(_ sender: Any) {
        showOrHideSkeletonButton.setTitle((view.isSkeletonActive ? "Show skeleton" : "Hide skeleton"), for: .normal)
        view.isSkeletonActive ? hideSkeleton() : showSkeleton()
    }
    
    @IBAction func transitionDurationStepperAction(_ sender: Any) {
        transitionDurationLabel.text = NSString(format:"Fade duration: %.2f s", transitionDurationStepper.value) as String
        if view.isSkeletonActive { showSkeleton() }
    }
    
     func showSkeleton() {
        if type.isGardient { showGradientSkeleton() }
        else { showSolidSkeleton() }
    }
    
    func hideSkeleton() {
        view.hideSkeleton(transition: SkeletonTransitionStyle(transitionDurationStepper.value))
    }
    
    func showSolidSkeleton() {
        if switchAnimated.isOn {
            view.showAnimatedSkeleton(usingColor: colorSelectedView.backgroundColor!, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        } else {
            view.showSkeleton(usingColor: colorSelectedView.backgroundColor!, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        }
    }
    
    func showGradientSkeleton() {
        let gradient = SkeletonGradient(baseColor: colorSelectedView.backgroundColor!, secondaryColor: colorSelectedView.tintColor)
        if switchAnimated.isOn {
            view.showAnimatedGradientSkeleton(usingGradient: gradient, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        } else {
            view.showGradientSkeleton(usingGradient: gradient, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        }
    }
    
    func showAlertPicker() {
        let alertView = UIAlertController(title: "Select color", message: "\n\n\n\n\n\n", preferredStyle: .alert)
        
        let pickerView = UIPickerView(frame: CGRect(x: 0, y: 50, width: 260, height: 115))
        pickerView.dataSource = self
        pickerView.delegate = self
        
        alertView.view.addSubview(pickerView)
        
        let action = UIAlertAction(title: "OK", style: .default) { [unowned pickerView, unowned self] _ in
            let row = pickerView.selectedRow(inComponent: 0)
            self.colorSelectedView.backgroundColor = Colors.color(row)
            if self.view.isSkeletonActive {
                self.hideSkeleton()
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                    self.showSkeleton()
                }
            }
        }
        alertView.addAction(action)
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        alertView.addAction(cancelAction)
        
        present(alertView, animated: false, completion: {
            pickerView.frame.size.width = alertView.view.frame.size.width
        })
    }
}

// MARK: - UIPickerViewDelegate, UIPickerViewDataSource

extension CollectionViewController: UIPickerViewDelegate, UIPickerViewDataSource {
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return Colors.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return Colors.title(row)
    }
}

// MARK: - UICollectionViewDelegateFlowLayout

extension CollectionViewController: UICollectionViewDelegateFlowLayout {
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: view.frame.width/3 - 10, height: view.frame.width/3 - 10)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return 5
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return 5
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return UIEdgeInsets(top: 0, left: 7, bottom: 0, right: 7)
    }
}

// MARK: - SkeletonCollectionViewDataSource

extension CollectionViewController: SkeletonCollectionViewDataSource {
    func collectionSkeletonView(_ skeletonView: UICollectionView,
                                supplementaryViewIdentifierOfKind: String,
                                at indexPath: IndexPath) -> ReusableCellIdentifier? {
        return nil
    }
    
    func numSections(in collectionSkeletonView: UICollectionView) -> Int { return 1 }
    
    func collectionSkeletonView(_ skeletonView: UICollectionView, cellIdentifierForItemAt indexPath: IndexPath) -> ReusableCellIdentifier {
        return "CollectionViewCell"
    }
    
    func collectionSkeletonView(_ skeletonView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 10
    }
    
    // MARK: - UICollectionViewDataSource
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 10
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "CollectionViewCell", for: indexPath) as! CollectionViewCell
        return cell
        
    }
}

